### PR TITLE
ISAICP-5024: Skip unchanged solutions by default.

### PR DESCRIPTION
**Pull request by idimopoulos**

### CODE REVIEW

File: web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+use Drupal\sparql_entity_storage\Database\Driver\sparql\ConnectionInterface;
+use Drupal\sparql_entity_storage\Entity\Query\Sparql\SparqlArg;
+use Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandlerInterface;
+use Drupal\sparql_entity_storage\SparqlEntityStorageGraphHandlerInterface;
+use EasyRdf\Sparql\Result;
+
+/**
+ * Helper class to generate a hash containing all data of an rdf entity.
+ *
+ * @package Drupal\joinup_federation
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: Sorry, it was auto generated by the IDE. Will remove.

File: web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+   * ordered by the field predicate and values alphabetically so that the end
+   * result cannot vary between same versions.
+   *
+   * @param array $entity_ids
+   *   An array of entity ids to generate the hashes for.
+   *
+   * @return array
+   *   An associative array of hashed indexed by their related entity id.
+   */
+  public function generateDataHash(array $entity_ids): array {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: I will give it a shot and will let you know.

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: From an example run, the following fields are not empty when converting the entity to an array:

Distributions : `id, rid, label, created, changed, uuid, path, graph, parent, field_ad_access_url, field_ad_description, field_ad_format, field_ad_licence`
Owner: `id, rid, uuid, path, graph, field_owner_name, field_owner_type`
Asset release: `id, rid, label, created, changed, uuid, path, graph, field_isr_contact_information, field_isr_description, field_isr_distribution, field_isr_is_version_of, field_isr_owner, field_isr_release_number`
Solution: `id, rid, label, created, changed, uuid, path, graph, field_is_contact_information, field_is_description, field_is_has_version, field_is_landing_page, field_is_language, field_is_owner, field_is_related_solutions, field_is_solution_type`

In the above examples, some ADMS fields are not present because it happens that the example file was not filled. The rest of the non-adms fields are filled with default values created along with the entity. Do you think it is worth using these values as well? that would make it a hash of the entity that is about to be created and not of the incoming data.

If we were to strip off the non ADMS fields then we either hardcode them or we will have to loop over the fields from the FieldHandler which makes it much less performant than querying for the data.

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: As for the integrity of the hash, I tried without diving deeper on the array and it turns that one distribution produced a different hash once so I had to change it to a ksort similar to the one I already have.
With that, I tried it multiple times and the results remain the same so we are covered in terms of that.

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: In general, I had something like
```
$return = [];
    foreach (Rdf::loadMultiple($entity_ids, ['staging']) as $entity_id => $entity) {
      $entity_as_array = $entity->toArray();
      $entity_as_array = array_filter($entity_as_array);

      foreach ($entity_as_array as &$values_array) {
        foreach ($values_array as &$values) {
          ksort($values);
        }
        ksort($values_array);
      }
      ksort($entity_as_array);


      $return[$entity_id] = md5(serialize($entity_as_array));
    }
    return $return;
```

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/SparqlEntityStorageTrait.php [hunk 3]
```diff
@@ -59,4 +59,33 @@ protected function getSparqlQuery(): SparqlQueryInterface {
...
+   * Returns the ids of solutions in the staging graph.
+   *
+   * @return array
+   *   And array of ids.
+   */
+  protected function getIncomingSolutionIds(): array {
+    return $this->getSparqlQuery()
+      ->graphs(['staging'])
+      ->condition('rid', 'solution')
+      ->notExists('field_is_version_of')
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/SparqlEntityStorageTrait.php]: You are absolutely right!

File: web/modules/custom/rdf_entity_provenance/config/install/field.storage.rdf_entity.field_provenance_entity_hash.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+langcode: en
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_entity_provenance/config/install/field.storage.rdf_entity.field_provenance_entity_hash.yml]: No reason really, it only seemed like a non provenance record property (our custom). I will switch to a base field.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php [hunk 5]
```diff
@@ -137,25 +136,19 @@ public function execute() {
...
-    if ($blacklist = array_values(array_diff($all_imported_ids, $this->whitelist))) {
-      $blacklist_ids = SparqlArg::serializeUris($blacklist, ' ');
-      $this->sparql->query("WITH <{$this->getGraphUri('sink')}> DELETE { ?entity_id ?p ?o } WHERE { ?entity_id ?p ?o . VALUES ?entity_id { {$blacklist_ids} } }");
+    $this->buildWhitelist($selected_solution_ids);
+    $all_imported_ids = $this->getAllIncomingIds();
+
+    // Remove the not selected entities, if any.
+    if ($not_selected = array_values(array_diff($all_imported_ids, $this->whitelist))) {
+      $not_selected_ids = SparqlArg::serializeUris($not_selected, ' ');
+      $this->sparql->query("WITH <{$this->getGraphUri('sink')}> DELETE { ?entity_id ?p ?o } WHERE { ?entity_id ?p ?o . VALUES ?entity_id { {$not_selected_ids} } }");
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php]: Actually, I didn't touch this line. This was how it was before. I think it was like that because this is the form submission of the UserSeleftionFilter and if there are too many solutions it would be slow to load all of them and then delete them (the `::deleteFromGraph` is receiving full entities), while this is blazing fast.

File: web/modules/custom/joinup_federation/tests/src/Kernel/UserSelectionFilterWhitelistBuilderTest.php [hunk 6]
```diff
@@ -1,173 +0,0 @@
-<?php
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/tests/src/Kernel/UserSelectionFilterWhitelistBuilderTest.php]: To be honest, the main reason is that it was too hard for me to write another one and compared to the effort needed, I didn't think it was worth more than the behat test.
The previous test was based on the UserSelectionFilter step and specifically the `::buildWhitelist`. The `::buildWhitelist` method was a recursive method running after a user was submitting the form, in a single run (not batched) and was calculating the entities whitelisted. The method was also not interacting with the pipeline or the batch storage (`::set/getBatchValue`) while building the list.

The execution method of the analyze step is mainly doing all of these.. That makes it hard to run standalone, without preparing a pipeline or in a UnitTest (this last one is actually a nightmare to me since I am not familiar with it).

On the other side, the new behat test that was extended, is checking trice the list of entities as shown in the UserSelectionFilter step, so after they are analyzed. All cases are checked fine with more complete data and run through the whole process of the importer.

On the bottom line, we were maintaining a list of custom solutions in the test, a duplicated array, that was supposed to check exactly what we see in the behat test as well. We do not assert that the hashes are the same of course, but we do assert the end result.

Still, the bottom line is that I initially tried, but all the dependencies, the pipeline, the previous step, the batch processing, they were quite a lot of effort for something we already test in behat.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php [hunk 7]
```diff
@@ -324,54 +228,19 @@ protected function getReferencedEntityIdsByBundle(EntityReferenceFieldItemListIn
    *   keyed by entity ID and having the entity labels as values.
    */
   protected function getEntitiesByCategory(): array {
-    $ids = $this->getSparqlQuery()
-      ->graphs(['staging'])
-      ->condition('rid', 'solution')
-      ->execute();
+    $this->ensureEntityDataLoaded();
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php]: Could we just use directly the persistent data store value?

File: web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/joinup_federation/src/JoinupFederationHashGenerator.php]: We've been decided that we keep the actual implementation as is much closer to "data from file".

File: web/modules/custom/joinup_federation/tests/src/Kernel/UserSelectionFilterWhitelistBuilderTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/joinup_federation/tests/src/Kernel/UserSelectionFilterWhitelistBuilderTest.php]: I added a test.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    $this->entityFieldManager = $entity_field_manager;
+    $this->provenanceHelper = $provenance_helper;
+    $this->rdfSchemaFieldValidator = $rdf_schema_field_validator;
+    $this->hashGenerator = $hash_generator;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: Do we really put that here? what about the type of the parameters? do we also add them in the inherited method from core? Should I update the interface as well?

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    $hash_ids = array_diff($ids, $this->getEntityIdsWithHashes());
+    $this->setEntityHashes($this->hashGenerator->generateDataHash($hash_ids));
+
+    // Handle the solutions and their dependencies of this iteration.
+    $storage = $this->getRdfStorage();
+
+    /** @var \Drupal\rdf_entity\RdfInterface[] $entities */
+    $entities = $storage->loadMultiple($solution_ids, ['staging']);
+    foreach ($entities as $id => $entity) {
+      if ($entity->bundle() === 'solution') {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: No you are right, that was because previously the full list of entities were going through this loop. Removed the condition.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+   *
+   * @param string $entity_id
+   *   The entity id.
+   *
+   * @return string
+   *   The hash related to the entity.
+   */
+  protected function getEntityHash(string $entity_id): string {
+    $this->ensureEntityDataLoaded();
+    return $this->entityHashes[$entity_id] ?? '';
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: The case would be when the user would request the hash of an entity before it is calculated. Generally, the two cases we have in our code is quite caseproof that it will always contain the entities. 
However, I do agree that it is better to simply return the hash and in case there is none, php will throw a warning/error about undefined index. Removing it.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+   * An associative array of hashes indexed by entity id.
+   *
+   * @var array
+   */
+  protected $entityHashes = NULL;
+
+  /**
+   * Loads the solution data from the persistent state.
+   */
+  protected function ensureEntityDataLoaded(): void {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: It will cause errors only if you do not run it before any processing. The difference is that it will have to be explicitly called every time a step needs any of the versions. Calling in the methods will ensure data are loaded. I can remove it if you want and explicitly call it in the two steps we are calling for it.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+   * An associative array indexed by the solution ids that contains the
+   * 'dependencies' and the 'category' values. The 'dependencies' is an
+   * associative array itself, of entity ids indexed by their bundle that are
+   * related to the solution. The 'category' is a string representing the status
+   * of the solution. Possible values are 'not_federated', 'federated',
+   * 'invalid_solution', 'federated_unchanged' and 'blacklisted'.
+   *
+   * @var array
+   */
+  protected $solutionData = NULL;
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: But I am not checking with `isset`, I am checking with `$this->solutionData === NULL`.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+        $this->buildSolutionDependencyTree($entity, $id);
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: The thing with the `::buildSolutionDependency` is that, unlike the previous implementation, the dependency list is a flat list. What I mean is that if a solution references a release and that release references a distribution, the dependency tree array will not be
```
array => (
   solution => array (
      release => array(distribution)
   ),
)
```
but
```
array => (
  solution => (
     release, distribution,
  ),
)
```
That is why the id that is being passed in the method remains as the parent so that if it descends further into the recursion, all entities are listed directly under the solution and not as a structured array.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+
+  /**
+   * Builds an array of categories for solutions stored in the dependency tree.
+   *
+   * @param array $solution_ids
+   *   A list of solution ids for which to calculate the category.
+   */
+  protected function buildSolutionsCategories(array $solution_ids): void {
+    foreach ($solution_ids as $parent) {
+      $provenance_record = $this->provenanceHelper->loadOrCreateEntityActivity($parent);
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: I was thinking that if a solution changes, then a change will most probably apply to the solution itself. For example, even if there is no change to the description, if a new release of the solution is created, then the `hasVersion` property of the solution will retrieve this release. Meaning that the solution hash will change anyway. So in real case scenarios, it will be rare that a solution will not change but a sub entity will. These would be cases like a contact information has changed or a distribution description is changed.
In all other cases, changes will probably propagate to the parent solution.
In most common cases, we avoid loading activities for all sub entities.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: Yes, but init to `NULL` is unusual. The best practice is to use `isset()`

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: On second thought, this is not relevant. I do not actually check whether the solution is already changed here. I thought I did. However, I am only passing entity ids around. I can load them and convert the `::isSolutionChanged` to something like `::containsChangedRecords` and pass a list of provenance records to compare against the current solution data.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+   *
+   * @param \Drupal\rdf_entity\RdfInterface $entity
+   *   The rdf entity currently checking for relations.
+   * @param string $parent
+   *   The solution parent. Set in the root of the tree.
+   */
+  protected function buildSolutionDependencyTree(RdfInterface $entity, string $parent): void {
+    // If this is the first time it is called, enter the solution id in the
+    // dependency tree.
+    if ($entity->id() === $parent && !$this->solutionDataRootExists($parent)) {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: Indeed I am already making sure to not descend into a solution in this recursive method. 
I removed the `::solutionDataRootExists` method and moved the initial setup of the root in the parent method. That way, it is ensured that it is called only once.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+      foreach ($this->getStagingReferencedEntities($field) as $referenced_entity) {
+        if ($this->hasSolutionDataChildDependency($parent, $referenced_entity)) {
+          // The entity has already been processed.
+          continue;
+        }
+
+        // Only for the case of the release, dive deeper and check whether there
+        // are more referenced entities. Normally, the only difference between
+        // the solution and the release, would be the distributions.
+        if ($referenced_entity->bundle() === 'asset_release') {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: In general, the importer wizard is strictly based in the ADMS V2 specs. Currently, excluding the collection, the only second or third level references are `solution > release > anything`.
We do not check for changes in the licence as we define the changes, so the only real reference is this. We are quite safe I think being based on the ontology for that. 

The other way would be to loop over all reference fields and go on the other side and do the opposite checks like `if (in_array($referenced_entity->bundle(), ['solution', 'licence']))` so either we check for the bundle to descend or we check for bundles to not descend.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+   * @return array
+   *   A flat list of dependencies concerning all solutions in the given list.
+   *   The flat list includes the solutions as well and adds the solutions and
+   *   the releases first in the list.
+   */
+  protected function getSolutionsWithDependenciesAsFlatList(array $solution_ids): array {
+    $this->ensureEntityDataLoaded();
+    $requested_dependencies = [];
+
+    $skipped_solutions = array_diff(array_keys($this->solutionData), $solution_ids);
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: Actually, the main idea is that if a solution has another solution as a dependency, it might be that the second will end up in the list of dependencies without asking for it. However, as we are not adding a solution as a dependency to another solution, this is now redundant.
Removing.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+    }
+    if ($this->entityHashes === NULL) {
+      $this->entityHashes = $this->hasPersistentDataValue('entity_hashes') ? $this->getPersistentDataValue('entity_hashes') : [];
+    }
+  }
+
+  /**
+   * Stores the entity data to the persistent pipeline state.
+   */
+  protected function storeEntityData(): void {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: Both of them were in the trait due to the `::ensureDataLoaded`method but I will move them to the step.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+  /**
+   * Returns whether a solution is on the root of the solution data.
+   *
+   * @param string $solution_id
+   *   The solution entity id.
+   *
+   * @return bool
+   *   Whether the solution id exists on the root level of the solution data.
+   */
+  protected function solutionDataRootExists(string $solution_id): bool {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: The method is removed as stated above.

File: web/modules/custom/joinup_federation/tests/modules/joinup_federation_test/src/Plugin/pipeline/Step/TestUserSelectionFilter.php [hunk 20]
```diff
@@ -20,10 +20,12 @@
 class TestUserSelectionFilter extends UserSelectionFilter {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/tests/modules/joinup_federation_test/src/Plugin/pipeline/Step/TestUserSelectionFilter.php]: No, not anymore.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ProvenanceActivity.php [hunk 21]
```diff
@@ -104,6 +105,16 @@ public function batchProcessIsCompleted() {
...
   public function execute() {
     $ids = $this->extractNextSubset('remaining_ids', static::BATCH_SIZE);
+
+    // Get all entities that are unchanged by fetching the dependencies of the
+    // unchanged solutions. That way, if any entity is a dependency to more than
+    // one solution (e.g. a contact information entity), out of which some
+    // solutions are unchanged and some are blacklisted, the entity itself
+    // should not be blacklisted as it is a dependency to some unchanged
+    // (federated) solutions.
+    $solutionIds = $this->getSolutionIdsMatchingCategory('federated_unchanged');
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ProvenanceActivity.php]: But why are we re-calculating them here? the method `::getSolutionIdsMatchingCategory` is simply accessing and tampering with the data already in the `solutionData` array.

The same applies for the next line, with the `::getSolutionsWithDependenciesAsFlatList` method.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php [hunk 22]
```diff
@@ -215,105 +213,11 @@ public function getAdditionalPersistentDataStore(FormStateInterface $form_state)
...
-      }, $field->getValue())));
-      if ($ids) {
-        /** @var \Drupal\rdf_entity\RdfInterface $entity */
-        foreach (Rdf::loadMultiple($ids, ['staging']) as $id => $entity) {
-          $return[$entity->bundle()][] = $id;
-        }
-      }
-    }
-    return $return;
+  protected function buildWhitelist(array $solution_ids): void {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php]: I mainly kept it for you to have a reference to the previous code but it was also in my mind to remove it if you agreed.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php (hunk 7 above)
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/UserSelectionFilter.php]: We do not store the label of the solution. Should we do so?

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: In general, that seemed a good idea when I started using it. Having a few methods only that were doing the major calculations like setting and retrieving basic properties. However, as I was working on the ticket, it turns out I needed too many methods to have this required in each one of them. Plus I ended up having methods required only by one step, living in the trait due to this method. This method is now renamed and called once in every step needed.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php]: Done, left without a default value and actually the `::ensureEntityDataLoaded` is not even checking for this anymore.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php [hunk 23]
```diff
@@ -0,0 +1,432 @@
...
+    /** @var \Drupal\rdf_entity\RdfInterface[] $solutions */
+    $solutions = $this->getRdfStorage()->loadMultiple($entity_ids, ['staging']);
+    $provenance_records = $this->provenanceHelper->loadOrCreateEntitiesActivity($entity_ids);
+
+    foreach ($solutions as $id => $solution) {
+      // Licences are stored in Joinup so changes do not affect incoming
+      // solution.
+      if ($solution->bundle() === 'licence') {
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php]: We are importing a solution and all related entities to a solution. However, we only select the solution on the UserSelection step. Due to that last part, a solution is changed if one of the following cases:
- The solution itself has changed.
- One of the related entities to the solution has changed.
Because, if one of the related entities changed, we need to import the whole solution tree (the solution and its dependencies) so that the changed entities are re-imported.

The PhpDoc is written properly and explains that this is not only solution ids. 
However, there is a problem with the name of the variable here. So I am renaming this.

**Comment by idimopoulos** [comment]
The previous build failed with a server side issue. I restarted the test. The previous failing build is https://app.continuousphp.com/git-hub/ec-europa/joinup-dev/build/19a18e51-d1e7-4c7e-bcd1-eaa333bd5f4b

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ThreeWayMerge.php [hunk 24]
```diff
@@ -144,7 +146,7 @@ public function execute() {
 
     /** @var \Drupal\rdf_entity\RdfInterface[] $incoming_entities */
     $incoming_entities = $incoming_ids ? $this->getRdfStorage()->loadMultiple($incoming_ids, ['staging']) : [];
-    $local_entities = $local_ids ? Rdf::loadMultiple($local_ids) : [];
+    $local_entities = $local_ids ? Rdf::loadMultiple($local_ids, ['default', 'draft']) : [];
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ThreeWayMerge.php]: What is the reason for this change? `['default', 'draft']` is default.

File: web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ThreeWayMerge.php [hunk 25]
```diff
@@ -115,6 +116,7 @@ public function initBatchProcess() {
     // Get the incoming entities that are stored also locally.
     $local_ids = $this->getSparqlQuery()
       ->condition('id', array_values($whitelist), 'IN')
+      ->graphs(['default', 'draft'])
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ThreeWayMerge.php]: Useless change. The query graphs are already defaulting to 'default' & 'draft'
